### ISSUE TITLE
Snippet for Twitter While Sharing Content is Not Yet Correct

### ISSUE THREAD

**Issue by ghost** [mecha-cms/extend.plugin.share#1]
I refer to this issue at https://github.com/mecha-cms/genome/issues/18 .

It is that the snippets for social networks are not yet correct. For example, the page description is taken on Twitter, but not the Twitter card.

The Twitter Card is a plugin that generates snippets to share content on Twitter.

If you want to share in Facebook and G +, you need to have thumbnails or whole pictures in the blog posts.

But the new Social Share plugin is already great :smile:

**Comment by taufik-nurrohman** [comment]
The tweet description is taken from the `text` parameter in URL. The maximum length is 140 characters. I have now updated the twitter button so that the `text` value will contains no more than 140 characters, including the URL and the space just before the URL.

Using [this snippet](http://mecha-cms.com/snippet/field-images) you can dynamicly create an `image` property to the page so that the image thumbnail will be included to the meta tag. Or, you can make it manually:

~~~ .yaml
---
title: Page Title
description: Page description.
author: Taufik Nurrohman
type: Markdown
image: path/to/image.png
...

Page content.
~~~

BTW, Mecha already has _Facebook Open Graph_ and _Twitter Cards_ plugin a long time ago. I just don&rsquo;t share it to Facebook because I think they are just complementaries.

 - http://mecha-cms.com/reference/extension/plugin/facebook-open-graph
 - http://mecha-cms.com/reference/extension/plugin/twitter-cards

**Comment by ghost** [comment]
Unfortunately, when I click on Twitter button, the blogpost description is still taken. If it is shorter than 140 characters, then it fits, but with some posts it is longer.

Could you please again click on the Twitter button with me somewhere and test it? But you do not have to share, just click on the button and see it.

Also, when I share on Google Plus, the URL is invalid.

Facebook Button should work now and also Pinterest works.

Thank you!

**Comment by taufik-nurrohman** [comment]
How about now?

**Comment by ghost** [comment]
You're just very good 👍 

Thank you for helping and fixing the problem.

All three social networks, such as FB, G + and Twitter, display real text snippets and URLs. So it is fitting.

**Comment by taufik-nurrohman** [comment]
But it still shorter than it should (still remains **83** characters). Hmmm.

![2017-04-27_211933](https://cloud.githubusercontent.com/assets/1669261/25487734/4813a3b2-2b8f-11e7-9006-e25b5e18485c.png)

What&rsquo;s wrong with this:

~~~ .php
// Maximum character(s) is 140
$s = substr($lot['description'], 0, 140);
// Minus the URL and a space
$s = substr($s, 0, 140 - strlen($lot['url']) - 1);
~~~

https://github.com/mecha-cms/extend.plugin.share/blob/8c002c5b8da6735b3482dea3276064a6b81051ea/share/lot/state/a.php#L15&L21

**Comment by ghost** [comment]
Very strange. I've now again checked the Facebook Object debugger and cleared the cache.

**Comment by ghost** [comment]
Can you please still look at the Pinterest button. The image from the blog post is not displayed correctly. I just shared this post(https://mechacms.wpzweinull.ch/article/mecha-v2-cms-social-share-buttons-installation) and see the result below on the screen.

![mecha-v2-cms-socialsharing-to-pinterest](https://cloud.githubusercontent.com/assets/7885632/25506393/a9712d48-2ba6-11e7-9ebb-74d522c132c7.png)


Thanks in advance.

**Comment by taufik-nurrohman** [comment]
Oh, it is because I was counting URL length that already encoded?

Please try to update again. Try right clicking on the image then choose copy image URL. Paste the URL here.

**Comment by ghost** [comment]
Thanks, but it still does not work.

I add my pictures mostly from https://internetblogger.de and use the URLs so.

The image URL is as follows: https://internetblogger.de/wp-content/uploads/2017/04/mecha-v2-cms-social-share-buttons-installieren.png

Whether it is SSL or something?

On the screenshot it looks like this. Now just in Pinterest.

![pinterest-mechav2cms-bug-social-share-buttons](https://cloud.githubusercontent.com/assets/7885632/25512111/a81218c0-2bcb-11e7-9678-512abe1fa192.png)

**Comment by taufik-nurrohman** [comment]
Oh, I think it is because the `alt` value on the image contains some `<p>` tags which breaks the `<img>` tag. I have stipped out all of the HTML tags contained in the `title` and `description` data so that they are now contain plain text only?

**Comment by ghost** [comment]
Hmm ... it still does not work.

I have just tried the Pinterest Image Debugger and he showed me these values.

Rich pins are not activated. What exactly does that mean?

![pinterest-rich-pins-enabled-error](https://cloud.githubusercontent.com/assets/7885632/25512636/8457bd3c-2bcf-11e7-90da-dfb043772478.png)

**Comment by taufik-nurrohman** [comment]
Sigh! So, then I have to make `rich-pins` plugin for Mecha 😭

https://developers.pinterest.com/docs/rich-pins/articles

But, it does not look different with Facebook&rsquo;s open graph meta tags?

**Comment by ghost** [comment]
Yes, it looks like this plugin with Rich Pins is needed to share in Pinterest.

I have just looked at the Articles Pins and it looks similar to the Facebook OpenGraph. You're right.
